File: erubit.platform/src/main/java/a/erubit/platform/interaction/AnalyticsManager.java
package a.erubit.platform.interaction;

import android.support.v4.app.Fragment;
import java.util.Iterator;
import java.util.ServiceLoader;

public class AnalyticsManager {
    private static final IAnalyticsReporter mReporterInstance = Container.mInstance;
    private static final AnalyticsManager mInstance = new AnalyticsManager();

    public static AnalyticsManager i() {
        return mInstance;
    }

    private AnalyticsManager() {
        initialize();
    }

    public void initialize() {
        mReporterInstance.initialize();
    }

    public void reportFragmentChanged(Fragment fragment) {
        if (fragment != null)
            mReporterInstance.report(fragment.getClass().getSimpleName(), "fragment");
        else
            mReporterInstance.report("navigationFragment", "fragment");
    }

    public void reportInteractionScreen() {
        mReporterInstance.report("Interaction", "WindowManager");
    }


    private static class Container
    {
        static final IAnalyticsReporter mInstance;

        static
        {
            IAnalyticsReporter instance;
            Iterator<IAnalyticsReporter> loader = ServiceLoader.load(IAnalyticsReporter.class).iterator();

            if (loader.hasNext()) {
                instance = loader.next();
            }
            else {
                instance = new EmptyAnalyticsReporter();
            }

            mInstance = instance;
        }
    }


    private static class EmptyAnalyticsReporter implements IAnalyticsReporter {
        @Override
        public void initialize() { }
        @Override
        public void report(String event, String category) { }
    }
}

File: erubit.platform/src/main/java/a/erubit/platform/interaction/AnalyticsManager.kt
package a.erubit.platform.interaction

import android.content.Context
import android.support.v4.app.Fragment
import java.util.*

class AnalyticsManager {
	fun initialize(context: Context) {
		mReporterInstance!!.initialize(context)
	}

	fun reportFragmentChanged(fragment: Fragment?) {
		mReporterInstance ?: return

		if (fragment != null)
			mReporterInstance.report(fragment.javaClass.simpleName, "fragment")
		else
			mReporterInstance.report("navigationFragment", "fragment")
	}

	fun reportInteractionScreen() {
		mReporterInstance ?: return
		mReporterInstance.report("Interaction", "WindowManager")
	}

	private object Container {
		var mInstance: IAnalyticsReporter? = null

		init {
			val instance: IAnalyticsReporter
			val loader: Iterator<IAnalyticsReporter> = ServiceLoader.load(IAnalyticsReporter::class.java).iterator()

			instance = if (loader.hasNext())
				loader.next()
			else
				EmptyAnalyticsReporter()

			mInstance = instance
		}
	}


	private class EmptyAnalyticsReporter : IAnalyticsReporter {
		override fun initialize(context: Context) {}
		override fun report(event: String, category: String) {}
	}


	companion object {
		private val mReporterInstance = Container.mInstance
		private val mInstance = AnalyticsManager()

		fun i(): AnalyticsManager {
			return mInstance
		}
	}
}
